src/resources/dac_esp32.cc

```diff
@@ -0,0 +1,240 @@
+// Copyright (C) 2022 Toitware ApS.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; version
+// 2.1 only.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// The license can be found in the file `LICENSE` in the top level
+// directory of this repository.
+
+#include "../top.h"
+
+#ifdef TOIT_FREERTOS
+
+#include <driver/gpio.h>
+#include <driver/dac.h>
+
+#include "../entropy_mixer.h"
+#include "../objects_inline.h"
+#include "../primitive.h"
+#include "../process.h"
+#include "../resource.h"
+#include "../resource_pool.h"
+#include "../vm.h"
+
+#include "../event_sources/gpio_esp32.h"
+#include "../event_sources/system_esp32.h"
+
+namespace toit {
+
+static constexpr int kDacMinFrequency = 130;
+static constexpr int kDacMaxFrequency = 5500;
+
+static constexpr dac_channel_t kInvalidChannel = static_cast<dac_channel_t>(-1);
+
+#ifdef CONFIG_IDF_TARGET_ESP32
+
+static dac_channel_t get_dac_channel(int pin) {
+  switch (pin) {
+    case 25: return DAC_CHANNEL_1;
+    case 26: return DAC_CHANNEL_2;
+    default: return kInvalidChannel;
+  }
+}
+
+#elif CONFIG_IDF_TARGET_ESP32S2
+
+static dac_channel_t get_dac_channel(int pin) {
+  switch (pin) {
+    case 17: return DAC_CHANNEL_1;
+    case 18: return DAC_CHANNEL_2;
+    default: return kInvalidChannel;
+  }
+}
+
+#elif CONFIG_IDF_TARGET_ESP32C3
+
+static dac_channel_t get_dac_channel(int pin) {
+  // The ESP32-C3 does not have any DAC.
+  return kInvalidChannel;
+}
+
+#elif CONFIG_IDF_TARGET_ESP32
+
+#error "Unsupported ESP32 target"
+
+#else
+
+static dac_channel_t get_dac_channel(int pin) {
+  return kInvalidChannel;
+}
+
+#endif
+
+class DacResourceGroup;
+
+static int _cosine_user_count = 0;
+
+class DacResource : public Resource {
+ public:
+  TAG(DacResource);
+  DacResource(DacResourceGroup* group, dac_channel_t channel);
+
+  virtual ~DacResource() override {
+    unuse_cosine();
+  }
+
+  dac_channel_t channel() const { return _channel; }
+
+  esp_err_t use_cosine() {
```

Comment:
I would probably not have these in the class definition (use_cosine / unuse_cosine). No need to have them inlined.